covid19-etl/etl/cityofchicago.py

```diff
@@ -0,0 +1,410 @@
+#  This ETL is for city of chicago dataset for COVID-19 Daily Cases, Deaths, and Hospitalizations (CDH)
+#  Reference: https://data.cityofchicago.org/Health-Human-Services/COVID-19-Daily-Cases-Deaths-and-Hospitalizations/naz8-j4nc
+
+import boto3
+import csv
+from datetime import datetime
+import time
+import json
+import requests
+import pandas as pd
+from contextlib import closing
+from utils.metadata_helper import MetadataHelper
+from utils.format_helper import (
+    derived_submitter_id,
+    format_submitter_id,
+)
+from etl import base
+import sys
+
+CITYOFCHICAGO_CDH_URL = "https://data.cityofchicago.org/resource/naz8-j4nc.csv"
```

Comment:
The dataset have data till the previous day, for example today is 04-26-2022 so it would have data till 04-25-2022. about the data looking weird from 2022-04-21 , they are not putting hospitalization data for some days. Their website have put out the information below.

> Because of the nature of data reporting to CDPH, hospitalizations will be blank for recent dates They will fill in on later updates when the data are received, although, as for cases and deaths, may continue to be updated as further data are received. 

([source](https://data.cityofchicago.org/Health-Human-Services/COVID-19-Daily-Cases-Deaths-and-Hospitalizations/naz8-j4nc/))

As the csv file would still have the column, currently the ETL would bypass this and put the value as zero. But we should come up with a solution when the hospitalization data is available. May I get suggestions on this?

Reply:
Mm the easiest (but not smartest lol) solution would be to resubmit the whole dataset every time, instead of querying the `latest_submitted_date` 🤔 
It sounds like they expect all 3 columns could be updated.
Maybe we could choose an arbitrary period of time to update? For example the `start_date` could be `latest_submitted_date - 1-2 month`?

Reply:
As the data is being updated everyday, ideal way that it should add data everyday. For this, We can do one thing, I can add a condition to check if hospitalization data is coming null or not in `parse_row` function. If its not null (meaning they started passing it again), I can raise exception and error it out. we can either wait for exception to occur and wait to get solution at that point or I can code right now that if this condition occur, I parse the file, check from which to which row the hospitalization records are empty and change `summary_group_demographics` record accordingly. Our main concern is to show `cases` and `deaths` data.

Reply:
I read `for cases and deaths, may continue to be updated` as "cases and deaths could be updated for past data". But i think you're right, the case and death data won't change once we submit it.

Reply:
> I parse the file, check from which to which row the hospitalization records are empty and change summary_group_demographics record accordingly

do you mean you would re-submit all the rows that have hospitalization data?

Reply:
Can we update the current records? If we can we can do it otherwise we can skip it for now and I can take it internally and come up with an answer what can we do about it.

Reply:
yes we can, that was my suggestion `resubmit the whole dataset every time`

but i'll be a lot of data. if we can make it smarter it'd be better. we could try querying all the records that have no hospitalization data, and only update those. That's also a pretty big query :s

Reply:
My approach for this would be, create a condition to check if date is greater than 21th April 2022 (the date where they stopped giving hospital data) and check if hospitalization column is empty or not, if not empty then take the records from file from 21th April 2022 to till date and make an api call to update hospitalization data on those records. The only query I would be making to sheepdog is to update those record.

Reply:
that's very manual 🤔 once the records are updated we have to remove the check. And when they stop adding hospitalization data again, we have to add a new date. I would say we can't be babysitting the ETL like that, it should be automatic

Reply:
Yeah That make sense. what appoarch do you think we can use to make it smarter? This idea was just something which came up to me. I can take time into this to think about the suitable way to make this happen.

Reply:
When the hospitalization data would be back, The datasource won't be updating the old columns (i.e. records from older dates) but only add the data in the new record (i.e. the current date) and ETL can work even if the record is present or empty. So I think we should be good with current process. Wydt?

Reply:
my suggestions from previous comments:

> the easiest (but not smartest lol) solution would be to resubmit the whole dataset every time

> we could try querying all the records that have no hospitalization data, and only update those

but now that i think about it, i think there's no way in Guppy to query data that does _not_ have a value

Something else you could do is use the `last_submission_identifier` to store which date the hospitalization data started being missing. That would work if all the missing values are in a row after a specific date, and not randomly distributed in the dataset

Reply:
So you are saying, I should defined `last_submission_identifier` as a field in `summary_group_demogrpahics`

Reply:
`last_submission_identifier` is a property in the project node

> That would work if all the missing values are in a row after a specific date, and not randomly distributed in the dataset

is that ^ the case?

Reply:
yes that's the case. its rows after specific date. Let me accommodate in the code

Reply:
ok then we could store in `last_submission_identifier` the number of the last row with hospitalization data

Reply:
@urvi-occ if you submit a record with a `submitter_id` that already exists, it will automatically do an update instead of a create. So in your case you would just need to process the row again and it should work

Reply:
oh. That make sense. Thank you

Reply:
I made a new commit where I am resubmitting missing data. so the data would be pulled from date `last_submission_identifier` till date. I also changed the date format and made it `%Y-%m-%d` . I'll be adding code for test in few days.